lib/rdf/changeset.rb

```diff
@@ -98,6 +99,24 @@ def readable?
       false
     end
 
+    ##
+    # Returns +false+ to indicate that this changeset _itself_ is not writable.
+    #
+    # @return [Boolean]
+    # @see    RDF::Writable#writable?
+    def writable?
+      false
+    end
+
+    ##
+    # Returns +false+ to indicate that this changeset _itself_ is not mutable.
+    #
```

Comment:
But, it seems that the Changeset _is_ mutable, as you can add or delete statements to it. It just doesn't implement the `RDF::Mutable` interface. An immutable Changeset would be something that was more frozen.

Reply:
Is this an issue with that docstring? I suppose I viewed it as an immutable shell with mutable contents. For instance you can freeze an `Array` or `Hash` and not be able to add or remove elements but you're perfectly free to go in and change the elements themselves. I personally think this is weird and would expect immutable to truly _mean_ immutable but I've only been working with Ruby in any serious way for a couple years and figured that the possibility for an immutable object with mutable contents was just how things were done.

At any rate I am not married to this statement and will be happy to change it.

Reply:
The question is not about documentation, but if it makes sense for Changeset to always return `#mutable?` as `false`, given that it does, in fact, allow statements to be added or deleted. The question is if returning `true` might imply that it implements the contract of `RDF::Mutable`, which I don't think it necessarily does.

IIRC (@no-reply was central to this), the idea was to separate `Dataset` from `Repository` so that a `Repository` could be considered as being based on an immutable `Dataset` with subsequent changes going through a `Changeset`. This could allow Moments-like chaining of Changesets based on transactions (`RDF::Transaction`) on an underlying Repository. Once a Changeset had been applied, it could conceivably live on as an immutable Changeset, providing for some form of rollback. The `RDF::Mutable#snapshot` interface is there to support this type of thing, but this doesn't seem to rely on Changeset, itself, which seems to be more of a transitional structure to keep track of changes as part of a transaction, which are ultimately applied to an underlying Repository.

Reply:
Empirically speaking, instances of `RDF::Changeset` _are_ "mutable", but the entire purpose of this exercise was to make them _not_ include `RDF::Mutable`. These predicates are here on @no-reply's [suggestion](https://github.com/ruby-rdf/rdf/issues/405#issuecomment-571263205) in #405. They could just as well be removed entirely.

When I made the analogous Perl module [`RDF::KV::Patch`](https://metacpan.org/pod/RDF::KV::Patch), the structure I created was likewise a union of two sets of statements (one of which, the statements to _remove_, permitted wildcards), but it was meaningless to consider them as a unitary bucket of statements, because one set of statements is strictly positive and the other is strictly negative. Indeed, I resorted to awkward method names like `add_this` and `dont_add_this` to add/remove statements to the "add" side, and `remove_this` and `dont_remove_this` to add/remove statements to the "remove" side.

Given that, I believe it would make great sense for the `inserts` and `deletes` _members_ of an `RDF::Changeset` instance to _themselves_ be `RDF::Mutable`, since then they could be manipulated directly, and the `insert` and `delete` methods of `RDF::Changeset` could be made proxies to `inserts.insert` and `deletes.insert`. (This assumes we're going to want to edit the contents of a changeset far less than simply appending to it, which is a safe bet.) The `inserts` side, moreover, can be made an in-memory RDF::Repository, which would also have the benefit of de-duplicating statements. The only question would be what to do about the `deletes` side.

Reply:
I don't think we need to spend too much time on mutability of a Changset; if it becomes an issue in the future, we can address it then.